File: dataStructures/excercises/lab6/java/prMaxiphobicHeap/src/dataStructures/heap/MaxiphobicHeap.java
/**
 * @author Pepe Gallardo, Data Structures, Grado en Informática. UMA.
 *
 * Maxiphobic Heaps
 */

package dataStructures.heap;

public class MaxiphobicHeap<T extends Comparable<? super T>> implements Heap<T> {

    // A node for an augmented binary tree
    private static class Node<E> {
        private E elem; // the element
        private int size; // the weight of tree rooted at this node
        private Node<E> left; // left child (null if no left child)
        private Node<E> right; // right child (null if no right child)
    }

    // Attribute for MaxiphobicHeap class
    private Node<T> root; // reference to root node of this Maxiphobic heap.
                          // null is heap is empty

    // Returns number of elements in tree stored at node
    private static int size(Node<?> node) {
        return node == null ? 0 : node.size;
    }

    // Merges two heaps. Returns merged heap.
    // Parameters are references to roots of heaps that should be merged.
    // Result should be a reference to root of resulting merged heap.
    private static <T extends Comparable<? super T>> Node<T> merge(Node<T> h1, Node<T> h2) {
        if (h1 == null) {
            return h2;
        }
        if (h2 == null) {
            return h1;
        }

        Node<T> root, key, child1, child2;
        if (h1.elem.compareTo(h2.elem) <= 0) {
            root = h1;
            key = h2;
            child1 = h1.left;
            child2 = h1.right;
        } else {
            root = h2;
            key = h1;
            child1 = h2.left;
            child2 = h2.right;
        }

        int child1Size = child1 == null ? 0 : child1.size;
        int child2Size = child2 == null ? 0 : child2.size;

        Node<T> n1, n2, n3;
        if (key.size > child1Size && key.size > child2Size) {
            n1 = key;
            n2 = child1;
            n3 = child2;
        } else if (child1Size > key.size && child1Size > child2Size) {
            n1 = child1;
            n2 = child2;
            n3 = key;
        } else {
            n1 = child2;
            n2 = child1;
            n3 = key;
        }
        root.left = n1;
        root.right = merge(n2, n3);

        return root;
    }

    // Constructor for MaxiphobicHeap class. Creates an empty Maxiphobic heap
    public MaxiphobicHeap() {
        root = null;
    }

    // Returns true if this Maxiphobic heap is empty
    public boolean isEmpty() {
        return root == null;
    }

    // Returns total number of elements in this Maxiphobic heap
    public int size() {
        return root == null ? 0 : root.size;
    }

    // Returns minimum element in this Maxiphobic heap
    public T minElem() {
        return root.elem;
    }

    // Removes minimum element from this Maxiphobic heap
    public void delMin() {
        if (isEmpty()) {
            throw new EmptyHeapException("delMin on empty Heap");
        } else {
            root = merge(root.left, root.right);
        }
    }

    // insert new element in this Maxiphobic heap
    public void insert(T elem) {
        Node<T> newHeap = new Node<T>();
        newHeap.elem = elem;
        newHeap.size = 1;
        newHeap.left = null;
        newHeap.right = null;

        root = merge(root, newHeap);
    }

    /**
     * Returns representation of this Maxiphobic heap as a String.
     */
    @Override
    public String toString() {
        String className = getClass().getSimpleName();
        StringBuilder sb = new StringBuilder();
        sb.append(className);
        sb.append("(");
        toStringRec(sb, root);
        sb.append(")");

        return sb.toString();
    }

    private static void toStringRec(StringBuilder sb, Node<?> node) {
        if (node == null) {
            sb.append("null");
        } else {
            sb.append("Node(");
            toStringRec(sb, node.left);
            sb.append(", ");
            sb.append(node.elem);
            sb.append(", ");
            toStringRec(sb, node.right);
            sb.append(")");
        }
    }
}

File: 1st semester/data-structures/excercises/lab6/haskell/DataStructures/Heap/MaxiphobicHeap.hs
-------------------------------------------------------------------------------
-- Maxiphobic Heaps
--
-- see: Fun with Binary Heap Trees
--      & Alternatives to two Classic Data Structures
--      by Chris Okasaki
--
-- Data Structures. Grado en Informática. UMA.
-- Pepe Gallardo, 2012
-------------------------------------------------------------------------------

module DataStructures.Heap.MaxiphobicHeap
  ( Heap
  , empty
  , isEmpty
  , minElem
  , delMin
  , insert
  , merge

  , mkHeap
  , size
  , isHeap

  , drawOnWith
  ) where

import DataStructures.Graphics.DrawTrees
import Test.QuickCheck

data Heap a  = Empty | Node a Int (Heap a) (Heap a) deriving Show

-- number of elements
size :: Heap a -> Int
size Empty            = 0
size (Node _ sz _ _)  = sz

empty :: Heap a
empty  = Empty

isEmpty :: Heap a -> Bool
isEmpty Empty  = True
isEmpty _      = False

-- Partially sorts heaps according to their sizes.
-- Returns largest heap in first position
sort3 :: Heap a -> Heap a -> Heap a -> (Heap a, Heap a, Heap a)
sort3 h1 h2 h3
    | x1>=x2 && x1>=x3 = (h1,h2,h3)
    | x2>=x1 && x2>=x3 = (h2,h1,h3)
    | otherwise = (h3,h2,h1)
    where
        x1 = size h1
        x2 = size h2
        x3 = size h3

-- Recursively merges smallest subheaps. Achieves O(log n) complexity
merge :: (Ord a) => Heap a -> Heap a -> Heap a
merge Empty h' = h'
merge h Empty = h
merge h1@(Node x1 w1 l1 r1) h2@(Node x2 w2 l2 r2)
    | x1 <= x2 = (Node x1 (w1+w2) heaviest1 (merge lightest1 lightest2))
    | otherwise = (Node x2 (w1+w2) heaviest2 (merge lightest3 lightest4))
    where
        (heaviest1, lightest1, lightest2) = sort3 (l1 r1 h2)
        (heaviest2, lightest3, lightest4) = sort3 (l2 r2 h1)

-- Returns a heap with a single element
singleton :: a -> Heap a
singleton x = Node x 1 Empty Empty 

-- Inserts an element in a heap
insert :: (Ord a) => a -> Heap a -> Heap a
insert x h = merge (singleton x) h 

-- Returns minimum element in heap
minElem :: Heap a -> a
minElem (Node x _ _ _) = x 

-- Deletes minimum element from heap
delMin :: (Ord a) => Heap a -> Heap a
delMin Empty = error "cannot delete min in empty heap"
delMin (Node _ _ lh rh) = merge lh rh



-- Efficient O(n) bottom-up construction for heaps
mkHeap :: (Ord a) => [a] -> Heap a
mkHeap []  = empty
mkHeap xs  = mergeLoop (map singleton xs)
  where
    mergeLoop [h]  = h
    mergeLoop hs   = mergeLoop (mergePairs hs)

    mergePairs []         = []
    mergePairs [h]        = [h]
    mergePairs (h:h':hs)  = merge h h' : mergePairs hs

-------------------------------------------------------------------------------
-- Generating arbritray Heaps
-------------------------------------------------------------------------------

instance (Ord a, Arbitrary a) => Arbitrary (Heap a) where
  arbitrary  = do
    xs <- arbitrary
    return (mkHeap xs)

-------------------------------------------------------------------------------
-- Invariants
-------------------------------------------------------------------------------

isHeap :: (Ord a) => Heap a -> Bool
isHeap Empty             = True
isHeap (Node x _ lh rh)  = x `lessEq` lh && x `lessEq` rh
                            && isHeap lh && isHeap rh
 where
  x `lessEq` Empty            = True
  x `lessEq` (Node x' _ _ _)  = x <= x'


-------------------------------------------------------------------------------
-- Drawing a Heap
-------------------------------------------------------------------------------

instance Subtrees (Heap a) where
  subtrees Empty             = []
  subtrees (Node _ _ lh rh)  = [lh,rh]

  isEmptyTree  = isEmpty

instance (Show a) => ShowNode (Heap a) where
  showNode (Node x _ _ _) = show x

drawOnWith :: FilePath -> (a -> String) -> Heap a -> IO ()
drawOnWith file toString = _drawOnWith file showHeap
 where
  showHeap (Node x _ _ _) = toString x
